test/xds/xds_client_ignore_resource_deletion_test.go

```diff
@@ -0,0 +1,474 @@
+//   - testResourceDeletionNotIgnored: When ignore_resource_deletion is unset,
+//     the xDSClient should delete the resource.
+//
+// Resource deletion is only applicable to Listener and Cluster resources.
+func (s) TestIgnoreResourceDeletionOnClient(t *testing.T) {
+	port1, cleanup := startTestService(t, nil)
+	t.Cleanup(cleanup)
+
+	port2, cleanup := startTestService(t, nil)
+	t.Cleanup(cleanup)
+
+	initialResourceOnServer := e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: []*listenerpb.Listener{e2e.DefaultClientListener(serviceName, rdsName)},
+		Routes:    []*routepb.RouteConfiguration{defaultRouteConfigWithTwoRoutes},
+		Clusters: []*clusterpb.Cluster{
+			defaultClientCluster(cdsName1, edsName1),
+			defaultClientCluster(cdsName2, edsName2),
+		},
+		Endpoints: []*endpointpb.ClusterLoadAssignment{
+			e2e.DefaultEndpoint(edsName1, "localhost", []uint32{port1}),
+			e2e.DefaultEndpoint(edsName2, "localhost", []uint32{port2}),
+		},
+		SkipValidation: true,
+	}
+
+	tests := []struct {
+		name           string
+		updateResource func(r *e2e.UpdateOptions)
+	}{
+		{
+			name: "listener",
+			updateResource: func(r *e2e.UpdateOptions) {
+				r.Listeners = nil
+			},
+		},
+		{
+			name: "cluster",
+			updateResource: func(r *e2e.UpdateOptions) {
+				r.Clusters = nil
+			},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("%s resource deletion ignored", test.name), func(t *testing.T) {
+			testResourceDeletionIgnored(t, initialResourceOnServer, test.updateResource)
+		})
+		t.Run(fmt.Sprintf("%s resource deletion not ignored", test.name), func(t *testing.T) {
+			testResourceDeletionNotIgnored(t, initialResourceOnServer, test.updateResource)
+		})
+	}
+}
+
+// This subtest tests the scenario where the bootstrap config has "ignore_resource_deletion"
+// set in "server_features" field. This subtest verifies that the resource was
+// not deleted by the xDSClient when a resource is missing the xDS response and
+// RPCs continue to succeed.
+func testResourceDeletionIgnored(t *testing.T, resources e2e.UpdateOptions, u func(r *e2e.UpdateOptions)) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	t.Cleanup(cancel)
+	ms := startManagementServer(t)
+	bootstrapContent := generateBootstrapContents(t, ms.Address, true)
+
+	// Update the management server with initial resources setup.
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	cResolver := generateCustomResolver(t, bootstrapContent)
+	cc, err := grpc.Dial(fmt.Sprintf("xds:///%s", serviceName), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(cResolver))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	verifyRPCtoAllEndpoints(t, cc)
+
+	// Mutate resource and update on the server.
+	u(&resources)
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Make RPCs for every 50ms for the next 500ms.
```

Comment:
Maybe add a little bit of detail to the comment? Something like:
```
// Make an RPC every 50ms for the next 500ms. This is to ensure that the updated resource
// is received from the management server and is processed by gRPC. Since
// resource deletions are ignored by the xDS client, we expect RPCs to all endpoints to keep succeeding.
```

Reply:
done.